rotkehlchen/exchanges/bybit.py

```diff
@@ -0,0 +1,535 @@
+
+            if response.status_code == HTTPStatus.TOO_MANY_REQUESTS:
+                if tries >= 1:
+                    backoff_seconds = 10 / tries
+                    log.debug(
+                        f'Got a 429 from Bybit. Backing off for {backoff_seconds}')
+                    gevent.sleep(backoff_seconds)
+                    tries -= 1
+                    continue
+
+                # else
+                raise RemoteError(
+                    f'Bybit api request for {response.url} failed with HTTP '
+                    f'status code {response.status_code} and response {response.text}',
+                )
+
+            if response.status_code != HTTPStatus.OK:
+                raise RemoteError(
+                    f'Bybit api request for {response.url} failed with HTTP status '
+                    f'code {response.status_code} and response {response.text}',
+                )
+
+            break  # else all good, we can break off the retry loop
+
+        log.debug(f'ByBit api response: {response.text}')
+        try:
+            json_ret: dict[str, Any] = json.loads(response.text)
+        except json.JSONDecodeError as e:
+            raise RemoteError('Bybit returned invalid JSON response') from e
+
+        if (result := json_ret.get('result')) is None:
+            raise RemoteError(f'Remote response is missing expected field result: {json_ret}')
+
+        if json_ret.get('retCode', 0) != 0 and json_ret.get('retMsg', 'OK') != 'OK':
+            raise RemoteError(f'Bybit returned error in request {json_ret}')
+
+        return result
+
+    def _paginated_api_query(
+            self,
+            endpoint: Literal[
+                'order/history',
+                'asset/deposit/query-record',
+                'asset/withdraw/query-record',
+            ],
+            options: dict[str, Any],
+    ) -> list[dict[str, Any]]:
+        if endpoint == 'order/history':
+            result_key = 'list'
+        else:
+            result_key = 'rows'
+
+        result = []
+        # copy since we are modifying the dict and could affect other queries using
+        # the same options
+        query_options = options.copy()
+        while True:
+            output = self._api_query(path=endpoint, options=query_options)
+            if (next_cursor := output.get('nextPageCursor')) is not None and len(next_cursor) != 0:
+                query_options['cursor'] = next_cursor
+
+            result.extend(output[result_key])
+            if len(output[result_key]) < query_options.get(result_key, PAGINATION_LIMIT):
+                break
+
+        return result
+
+    def query_online_trade_history(
+            self,
+            start_ts: Timestamp,
+            end_ts: Timestamp,
+    ) -> tuple[list[Trade], tuple[Timestamp, Timestamp]]:
+        """
+        Query trades from bybit in the spot category.
+        The API is limited to query trades up to two years into the past. We can query time ranges
```

Comment:
Is `start_ts` changed every time you query? So does it remember the last period you queried and the end ts of the last period become the new start ts? If yes then good. If not this need to happen, otherwise every time we got ino history querying you would get to query 2 years of history

Reply:
No it doesn't query everything every time. We have the cache and we query only what remains https://github.com/rotki/rotki/blob/9650dd0ad8050ca766c4b36d688047898a93e583/rotkehlchen/exchanges/exchange.py#L266